### PR TITLE
Library Panels: Improve unsaved changes handling

### PR DESCRIPTION
**Pull request by kaydelaney**
**What this PR does / why we need it**:
Improves handling of unsaved changes so the user is prompted with a modal dialogue if they press esc, the back arrow component, or the browser's back button should they have any unsaved changes.

### CODE REVIEW

File: public/app/features/dashboard/components/PanelEditor/PanelEditor.tsx [hunk 1]
```diff
@@ -117,15 +118,20 @@ export class PanelEditorUnconnected extends PureComponent<Props> {
...
     });
   };
 
+  onDiscard = () => {
+    this.props.setDiscardChanges(true);
+    this.props.panel.hasChanged = false;
+
+    this.onBack();
+  };
+
```
Comment (hugohaggmark) [inline public/app/features/dashboard/components/PanelEditor/PanelEditor.tsx:134]: This code seems very similar to the code in the `exitPanelEditor` Thunk, is there a way we can move this into a separate thunk that can be called from here and from the `exitPanelEditor`.

**Review by hugohaggmark — COMMENTED** [review]
Looks good and works as expected. I have some comments:
- The order of the buttons in the confirm differs from the order when saving a dashboard
image
image

- When I save a Library Panel in edit mode, I think it would be great to get some feedback like a notification, maybe?
- Seems like changing Angular Props in the Graph doesn't trigger the change detection, not sure this is an additional issue though?

File: public/app/features/dashboard/components/PanelEditor/PanelEditor.tsx (hunk 1 above)
Comment (kaydelaney) [reply public/app/features/dashboard/components/PanelEditor/PanelEditor.tsx:134]: Good idea!

**Comment by kaydelaney** [comment]
> Seems like changing Angular Props in the Graph doesn't trigger the change detection, not sure this is an additional issue though?

I think this is just a consequence of our change from using diffing to detect changes to using a dirty flag. Something to investigate though, I agree.

**Review by hugohaggmark — APPROVED** [review]
Looks good 🎉

**Comment by hugohaggmark** [comment]
Any reason why this isn't merged yet?

**Pull request merged**